### ISSUE TITLE
How to remove Child action cache

### ISSUE THREAD

**Issue by LockTar** [moonpyk/mvcdonutcaching#36]
I've searched for this but I can't find it. See here an example:

Index action of the HomeController (homepage):
```c#
[AllowAnonymous]
public ActionResult Index()
{
   return View();
}
```

ChildAction of the NewsController: 
```c#
[AllowAnonymous]
[ChildActionOnly]
[DonutOutputCache(Duration = 600, Location = OutputCacheLocation.Server)]
public PartialViewResult LastArticles(int numberOfArticles)
{
	return PartialView("_LastArticles", db.NewsArticles
	.Include(i => i.Tags)
	.Include(i => i.SeoTags)
	.Where(n => n.Status == PublishStatus.Published)
	.OrderByDescending(n => n.PublishDate)
	.Take(numberOfArticles)
	.ToList());
}
```

Index view of the HomeController:
```asp
@{ Html.RenderAction("LastArticles", "News", new { numberOfArticles = 2 }); }
```

To clear the cache I have an **Admin** area in my application with a controller and action in it to update the data stored by the child action. So when a news article is updated. The cache should get refreshed on the homepage.

In that action I have the following code:
```c#
var cacheManager = new OutputCacheManager();
cacheManager.RemoveItem("News", "LastArticles", new { area = "", numberOfArticles = 2 });
cacheManager.RemoveItem("News", "LastArticles", new { area = "" });
```

I tried multiple versions but without luck. Can you help me?

**Comment by LockTar** [comment]
Finally found the solution. The **area** is ignored in the url for removing the child action. This is different then the default OutputCache attribute. I think this is a bug?

So the solution is:
```c#
cacheManager.RemoveItem("News", "LastArticles", new { numberOfArticles = 2 });
```

See for more information this [StackOverlow question](http://stackoverflow.com/questions/28061486/mvcdonutcaching-how-to-remove-child-action-cache-mvc-donut-caching).

**Comment by moonpyk** [comment]
Thanks for reporting, I know exactly what happens.
When generating the cache key during action invocation, the default keybuilder only builds an area key component if the area is not null or empty.
By calling RemoveItem[s] with an area component set to empty, the area key component shouldn't be built either, otherwise no key will match for removal

**Comment by LockTar** [comment]
Thank you for your reply.
So if I understand it right, this is by design? Does this effect other NewsControllers with Action LastArticles in other Area's?

**Comment by moonpyk** [comment]
Nope this is a bug, the area part should be ignored if empty for removal action.
Thus, the normal behaviour would be to remove the area impacted keys. This is what you are doing by calling cacheManager.RemoveItem("News", "LastArticles", new { numberOfArticles = 2 });

Generally speaking it's a bad idea to have controllers with the same name in different areas, we got problems in the past with that.
The solution we adopted was to prefix our area controllers with the area name itself, making url routing tweaks afterwards to hide the redundancy. 

eg NewsController and AdminNewsController in the Admin area + route tweaks to avoid /Admin/AdminNews and have /Admin/News routes instead.
